tokio-threadpool/src/lib.rs

```diff
@@ -0,0 +1,1798 @@
+                continue;
+            }
+
+            // Starting to get sleeeeepy
+            if spin_cnt < 32 {
+                spin_cnt += 1;
+
+                // Don't do anything further
+            } else if spin_cnt < 256 {
+                spin_cnt += 1;
+
+                // Yield the thread
+                thread::yield_now();
+            } else {
+                if !self.sleep() {
+                    return;
+                }
+            }
+
+            // If there still isn't any work to do, shutdown the worker?
+        }
+
+        self.should_finalize.set(true);
+    }
+
+    /// Checks the worker's current state, updating it as needed.
+    ///
+    /// Returns `true` if the worker should run.
+    #[inline]
+    fn check_run_state(&self, first: bool) -> bool {
+        let mut state: WorkerState = self.entry().state.load(Acquire).into();
+
+        loop {
+            let pool_state: State = self.inner.state.load(Acquire).into();
+
+            if pool_state.is_terminated() {
+                return false;
+            }
+
+            let mut next = state;
+
+            match state.lifecycle() {
+                WORKER_RUNNING => break,
+                WORKER_NOTIFIED | WORKER_SIGNALED => {
+                    // transition back to running
+                    next.set_lifecycle(WORKER_RUNNING);
+                }
+                lifecycle => panic!("unexpected worker state; lifecycle={}", lifecycle),
+            }
+
+            let actual = self.entry().state.compare_and_swap(
+                state.into(), next.into(), AcqRel).into();
+
+            if actual == state {
+                break;
+            }
+
+            state = actual;
+        }
+
+        // If this is the first iteration of the worker loop, then the state can
+        // be signaled.
+        if !first && state.is_signaled() {
+            trace!("Worker::check_run_state; delegate signal");
+            // This worker is not ready to be signaled, so delegate the signal
+            // to another worker.
+            self.inner.signal_work(&self.inner);
+        }
+
+        true
+    }
+
+    /// Runs the next task on this worker's queue.
+    ///
+    /// Returns `true` if work was found.
+    #[inline]
+    fn try_run_task(&self, notify: &Arc<Notifier>) -> bool {
+        use coco::deque::Steal::*;
+
+        // Poll the internal queue for a task to run
+        match self.entry().deque.steal_weak() {
+            Data(task) => {
+                self.run_task(task, notify);
+                true
+            }
+            Empty => false,
+            Inconsistent => true,
+        }
+    }
+
+    /// Tries to steal a task from another worker.
+    ///
+    /// Returns `true` if work was found
+    #[inline]
+    fn try_steal_task(&self, notify: &Arc<Notifier>) -> bool {
+        use coco::deque::Steal::*;
+
+        let len = self.inner.workers.len();
+
+        let mut found_work = false;
+
+        worker_loop!(idx = len; len; {
```

Comment:
Shouldn't this be stealing from a random, active worker? This seems to be going through the workers in order.

Reply:
Hmm... you are correct. It was supposed to start at a random index but that seems to have gotten lost / forgotten! Thanks.

Reply:
On mobile ATM and didn't inspect the loop mentioned, but are we confident that starting at a random index is sufficiently random?  Something tells me we actually need to randomly index into a proper set of available workers, otherwise there may be statistical bias towards picking higher numbered workers

Reply:
It works as a ring, so there isn't really a "higher" index worker.

IIRC, this bit was inspired from Java's ForkJoin pool.